src/lib/homestore_backend/gc_manager.cpp

```diff
@@ -0,0 +1,389 @@
+    } else {
+        LOGINFO("pdev gc actor already exists for pdev_id: {}", pdev_id);
+    }
+    return it->second;
+}
+
+std::shared_ptr< GCManager::pdev_gc_actor > GCManager::get_pdev_gc_actor(uint32_t pdev_id) {
+    auto it = m_pdev_gc_actors.find(pdev_id);
+    if (it == m_pdev_gc_actors.end()) {
+        LOGERROR("pdev gc actor not found for pdev_id: {}", pdev_id);
+        return nullptr;
+    }
+    return it->second;
+}
+
+bool GCManager::is_eligible_for_gc(std::shared_ptr< HeapChunkSelector::ExtendedVChunk > chunk) {
+    if (!chunk->m_pg_id.has_value()) return false;
+    auto defrag_blk_num = chunk->get_defrag_nblks();
+    auto total_blk_num = chunk->get_total_blks();
+
+    // defrag_blk_num > (GC_THRESHOLD_PERCENT/100) * total_blk_num, to avoid floating point number calculation
+    // TODO: avoid overflow here.
+    return 100 * defrag_blk_num > total_blk_num * GC_GARBAGE_RATE_THRESHOLD;
+}
+
+void GCManager::scan_chunks_for_gc() {
+    for (const auto& [_, chunk] : m_chunk_selector->get_all_chunks()) {
+        if (is_eligible_for_gc(chunk)) {
+            auto pdev_id = chunk->get_pdev_id();
+            auto it = m_pdev_gc_actors.find(pdev_id);
+            if (it != m_pdev_gc_actors.end()) {
+                auto& actor = it->second;
+                auto [promise, _ /* future */] = folly::makePromiseContract< bool >();
+                // TODO:: use the future to check if the normal task is completed when necessary
+                actor->add_gc_task(
+                    TASK_PRIORITY::NORMAL,
+                    {chunk->get_chunk_id(), static_cast< uint8_t >(TASK_PRIORITY::NORMAL), std ::move(promise)});
+            }
+        }
+    }
+}
+
+/* pdev_gc_actor */
+
+GCManager::pdev_gc_actor::pdev_gc_actor(uint32_t pdev_id, std::shared_ptr< HeapChunkSelector > chunk_selector,
+                                        std::shared_ptr< GCBlobIndexTable > index_table, HSHomeObject* homeobject) :
+        m_pdev_id{pdev_id},
+        m_chunk_selector{chunk_selector},
+        // we have a queue for each priority, so the size is the same as the number of priorities
+        m_gc_task_queue{static_cast< size_t >(TASK_PRIORITY::PRIORITY_COUNT)},
+        m_reserved_chunk_queue{RESERVED_CHUNK_NUM_PER_PDEV},
+        m_index_table{index_table},
+        m_hs_home_object{homeobject} {
+
+    RELEASE_ASSERT(index_table, "index_table for a gc_actor should not be nullptr!!!");
+}
+
+void GCManager::pdev_gc_actor::start() {
+    bool stopped = true;
+    if (!m_is_stopped.compare_exchange_strong(stopped, false, std::memory_order_release, std::memory_order_relaxed)) {
+        LOGERROR("pdev gc actor for pdev_id={} is already started, no need to start again!", m_pdev_id);
+        return;
+    }
+    // thread number is the same as reserved chunk, which can make sure every gc thread can take a reserved chunk
+    // for gc
+    m_gc_executor = std::make_shared< folly::IOThreadPoolExecutor >(RESERVED_CHUNK_NUM_PER_PDEV);
```

Comment:
this implementation is wired,  as it is using `folly::IOThreadPoolExecutor` there is no point to keep a busy loop.
We can simply active an executor in both
1.  `scan_chunks_for_gc` where the topK  normal GC candidates are picked.
2.   in the emergent GC handler also add a task into the executor.
by doing 
```
  m_gc_executor->add([this] {
                GCManager::gc_task task;
                if (m_gc_task_queue.try_dequeue(task)) {
                    process_gc_task(task);
                } 
}
```

Reply:
Even more, the task queue can be removed as its only functionality is multiplexing emergent_GC(only one) and normal_gc(K). 

Assuming RESERVED_CHUNK_NUM_PER_PDEV = K,  we can get K-1 chunks for normal GC and always reserve one chunk for emergent GC.

Reply:
Good point on multiple emergent GC.

This is a chicken-egg problem :  If we choose to put all GC candidate into a Q, then we need a priority queue to prioritize Emergent GC, then the executor(threads) need to pulling on the Q,  it is a small optimization to use CV to avoid busy loop.

However,  if the design changed from very beginning, that do not put all GC candidate into Q,  just enough tasks to keep threads busy( meaning keeping only K submitted task  where K <= thread-num) .  Then we dont need to maintain priority as there will be no pending normal GC in the Q.  In this case, the task-Q built into folly thread pool is enough.